### PR TITLE
TlsCredentials: Comparator implementation

### PR DESCRIPTION
**Pull request by yashykt**
This PR allows channels configured with the same/similar `TlsCredentials` objects to use the same underlying subchannels. To avoid sharing subchannels with other channels, the channel arg `GRPC_ARG_USE_LOCAL_SUBCHANNEL_POOL` can be set to true.

Since `XdsCredentials` objects use `TlsCredentials` underneath, this would also prevent recreation of subchannels every time there's an xDS update.

Non-`Release Notes` Notes -
- Adds comparator methods for `grpc_tls_certificate_verifier`, `grpc_tls_certificate_provider` and an `operator==` method `grpc_tls_credentials_options`. (The comparator for `grpc_tls_certificate_provider` was not needed to achieve the reuse of subchannels at this time, but added it in the interest of consistency)
- To also be sure that any fields added to `grpc_tls_credentials_options` also result in an update to `operator==` method, added a generator script for `grpc_tls_credentials_options.h` and `grpc_tls_credentials_options_comparator_test.cc`

### CODE REVIEW

**Comment by yihuazhang** [comment]
@ZhenLian, could you please review this PR? It make changes to advanced TLS.

**Comment by ZhenLian** [comment]
> @ZhenLian, could you please review this PR? It make changes to advanced TLS.

Sure, will do!

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 1]
```diff
@@ -52,6 +53,25 @@ struct grpc_tls_certificate_verifier
   // Operations that will be performed when a request is cancelled.
   // This is only needed when in async mode.
   virtual void Cancel(grpc_tls_custom_verification_check_request* request) = 0;
+
+  // Compares this grpc_tls_certificate_verifier object with \a other.
+  // If this method returns 0, it means that gRPC can treat the two channel
+  // credentials as effectively the same.
+  int cmp(const grpc_tls_certificate_verifier* other) const {
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:60]: Please call this `Compare()`, and the internal method `CompareImpl()`.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc [hunk 2]
```diff
@@ -30,6 +30,15 @@
 
 namespace grpc_core {
 
+namespace {
+const char kTlsCertificateVerifierTypeExternal[] = "external";
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc:34]: Suggest making this a private static member of the `ExternalCertificateVerifier` class, so that it's not possible for an external impl to accidentally use it.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc [hunk 3]
```diff
@@ -80,6 +89,12 @@ void ExternalCertificateVerifier::OnVerifyDone(
   }
 }
 
+namespace {
+const char kTlsCertificateVerifierTypeHostName[] = "hostname";
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc:93]: Same here, private static within `HostNameCertificateVerifier`.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 4]
```diff
@@ -37,7 +38,7 @@
 struct grpc_tls_certificate_verifier
     : public grpc_core::RefCounted<grpc_tls_certificate_verifier> {
  public:
-  grpc_tls_certificate_verifier() = default;
+  explicit grpc_tls_certificate_verifier(const char* type) : type_(type) {}
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:41]: Please add the same sort of comment as in #28973.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc [hunk 5]
```diff
@@ -80,6 +89,12 @@ void ExternalCertificateVerifier::OnVerifyDone(
...
+}  // namespace
+
+HostNameCertificateVerifier::HostNameCertificateVerifier()
+    : grpc_tls_certificate_verifier(kTlsCertificateVerifierTypeHostName) {}
 bool HostNameCertificateVerifier::Verify(
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc:98]: Please add a blank line before this one.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h [hunk 6]
```diff
@@ -126,6 +126,25 @@ struct grpc_tls_credentials_options
...
+  bool operator==(const grpc_tls_credentials_options& other) const {
+    return cert_request_type_ == other.cert_request_type_ &&
+           verify_server_cert_ == other.verify_server_cert_ &&
+           min_tls_version_ == other.min_tls_version_ &&
+           max_tls_version_ == other.max_tls_version_ &&
+           (verifier_ == other.verifier_ ||
+            (verifier_ != nullptr && other.verifier_ != nullptr &&
+             verifier_->cmp(other.verifier_.get()) == 0)) &&
+           check_call_host_ == other.check_call_host_ &&
+           provider_ == other.provider_ &&
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:138]: Why are we plumbing the `cmp()` method into the verifier but not into the provider?  Seems like we should be consistent here: either we always require both to be the same instance, or we require both to compare the same.

I'd be fine with either approach.  Requiring them to be the same instance is simpler, because it eliminates the need to plumb the `cmp()` method into the verifier API.  But I'd also be fine with plumbing the `cmp()` method into the provider as well.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h [hunk 7]
```diff
@@ -126,6 +126,25 @@ struct grpc_tls_credentials_options
   // not enable CRL checking. Only supported for OpenSSL version > 1.1.
   void set_crl_directory(std::string path) { crl_directory_ = std::move(path); }
 
+  bool operator==(const grpc_tls_credentials_options& other) const {
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:129]: One down-side of this approach is that any time we add a new field to `grpc_tls_credentials_options`, we need to remember to change `operator==()` to look at the new field.  Is there some way of automating this such that that happens automatically?  If not, we could be opening ourselves up to a security problem in the future.

File: src/core/lib/security/credentials/xds/xds_credentials.cc [hunk 8]
```diff
@@ -61,44 +61,57 @@ bool XdsVerifySubjectAlternativeNames(
...
-        !XdsVerifySubjectAlternativeNames(
-            request->peer_info.san_names.dns_names,
-            request->peer_info.san_names.dns_names_size,
-            xds_certificate_provider_->GetSanMatchers(cluster_name_))) {
-      *sync_status = absl::Status(
-          absl::StatusCode::kUnauthenticated,
-          "SANs from certificate did not match SANs from xDS control plane");
-    }
-    return true; /* synchronous check */
+const char kTlsCertificateVerifierTypeXds[] = "xds";
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.cc:64]: Suggest making this private static within `XdsCertificateVerifier`.

File: src/core/lib/security/credentials/xds/xds_credentials.cc [hunk 9]
```diff
@@ -61,44 +61,57 @@ bool XdsVerifySubjectAlternativeNames(
...
-  std::string cluster_name_;
-};
+void XdsCertificateVerifier::Cancel(
+    grpc_tls_custom_verification_check_request*) {}
 
-}  // namespace
+int XdsCertificateVerifier::cmp_impl(
+    const grpc_tls_certificate_verifier* other) const {
+  auto* o = static_cast<const XdsCertificateVerifier*>(other);
+  if (xds_certificate_provider_ == o->xds_certificate_provider_ &&
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.cc:108]: If we plumb the `cmp()` method into the provider, then we should probably use it here.

File: src/core/lib/security/credentials/xds/xds_credentials.cc [hunk 10]
```diff
@@ -61,44 +61,57 @@ bool XdsVerifySubjectAlternativeNames(
...
+      cluster_name_ == o->cluster_name_) {
+    return 0;
+  }
+  return QsortCompare(static_cast<const grpc_tls_certificate_verifier*>(this),
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.cc:112]: I don't think we need to fall back on comparing the pointer values if we're comparing all of the data members anyway.  Instead, I think we can do something like this:

```
int r = QsortCompare(xds_certificate_provider_, o->xds_certificate_provider_);
if (r != 0) return r;
return cluster_name_.compare(o->cluster_name_);
```

File: test/core/security/credentials_test.cc [hunk 11]
```diff
@@ -3677,6 +3613,62 @@ TEST(CredentialsTest,
...
+  options_2->set_certificate_verifier(
+      MakeRefCounted<HostNameCertificateVerifier>());
+  auto* tls_creds_2 = grpc_tls_credentials_create(options_2);
+  EXPECT_EQ(tls_creds_1->cmp(tls_creds_2), 0);
+  EXPECT_EQ(tls_creds_2->cmp(tls_creds_1), 0);
+  grpc_channel_credentials_release(tls_creds_1);
+  grpc_channel_credentials_release(tls_creds_2);
+}
+
+TEST(CredentialsTest, TestTlsCredentialsCompareFailure) {
```
Comment (markdroth) [inline test/core/security/credentials_test.cc]: To be actually safe here, we need a test that shows that the comparison fails if any one field is different.  Otherwise, we could be subject to serious security bugs.

And this is another thing that would be really nice to automatically update whenever we add a field to `grpc_tls_credentials_options`.

File: test/core/security/credentials_test.cc [hunk 12]
```diff
@@ -3701,6 +3693,72 @@ TEST(CredentialsTest, TestXdsCredentialsCompareFailure) {
   grpc_channel_credentials_release(xds_creds_2);
 }
 
+TEST(TlsCertificateVerifierTest, ComparingDifferentObjectTypesFails) {
```
Comment (markdroth) [inline test/core/security/credentials_test.cc]: The verifier tests belong in test/core/security/grpc_tls_certificate_verifier_test.cc.

File: test/core/security/credentials_test.cc [hunk 13]
```diff
@@ -3701,6 +3693,72 @@ TEST(CredentialsTest, TestXdsCredentialsCompareFailure) {
...
+                                                       nullptr, nullptr};
+  ExternalCertificateVerifier external_verifier_1(&verifier_1);
+  grpc_tls_certificate_verifier_external verifier_2 = {nullptr, nullptr,
+                                                       nullptr, nullptr};
+  ExternalCertificateVerifier external_verifier_2(&verifier_2);
+  EXPECT_NE(external_verifier_1.cmp(&external_verifier_2), 0);
+  EXPECT_NE(external_verifier_2.cmp(&external_verifier_1), 0);
+}
+
+TEST(TlsCertificateVerifierTest, XdsCertificateVerifierSuccess) {
```
Comment (markdroth) [inline test/core/security/credentials_test.cc]: The xDS verifier tests should probably live in test/core/security/xds_credentials_test.cc.

File: test/core/security/credentials_test.cc [hunk 14]
```diff
@@ -56,24 +56,28 @@
...
 #include "src/core/lib/security/credentials/oauth2/oauth2_credentials.h"
+#include "src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h"
+#include "src/core/lib/security/credentials/xds/xds_credentials.h"
 #include "src/core/lib/security/transport/auth_filters.h"
 #include "src/core/lib/uri/uri_parser.h"
 #include "test/core/util/test_config.h"
 
-using grpc_core::internal::grpc_flush_cached_google_default_credentials;
-using grpc_core::internal::set_gce_tenancy_checker_for_testing;
+namespace grpc_core {
```
Comment (markdroth) [inline test/core/security/credentials_test.cc]: The namespace changes here are definitely desirable, but they should probably go in a separate PR.

File: test/cpp/end2end/xds/xds_end2end_test.cc [hunk 15]
```diff
@@ -7538,13 +7538,11 @@ class XdsSecurityTest : public BasicTest {
     constexpr int kRetryCount = 100;
     int num_tries = 0;
     for (; num_tries < kRetryCount; num_tries++) {
-      // Give some time for the updates to propagate.
-      gpr_sleep_until(grpc_timeout_milliseconds_to_deadline(100));
+      // Restart the servers to force a reconnection so that previously
```
Comment (markdroth) [inline test/cpp/end2end/xds/xds_end2end_test.cc:7541]: Can you explain why this change is here?  I imagine it has something to do with the following TODO, but the exact reasoning isn't clear.

https://github.com/grpc/grpc/blob/87068e02c3c4d34f51019b3ccdceea1230c92fea/src/core/lib/security/credentials/xds/xds_credentials.cc#L167

Speaking of which, shouldn't we remove that TODO as part of this PR?

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!

Please let me know if you have any questions.  Thanks!

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 7 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:129]: We get default comparison operations in C++20 :)

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 6 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:138]: We need it in verifier because of an explicit need in XdsCertificateVerifier https://github.com/grpc/grpc/blob/87068e02c3c4d34f51019b3ccdceea1230c92fea/src/core/lib/security/credentials/xds/xds_credentials.cc#L163.. We don't have such a need for providers yet.

File: src/core/lib/security/credentials/xds/xds_credentials.cc (hunk 9 above)
Comment (yashykt) [reply src/core/lib/security/credentials/xds/xds_credentials.cc:108]: As mentioned earlier, there isn't a need for such an operation on providers right now

File: test/cpp/end2end/xds/xds_end2end_test.cc (hunk 15 above)
Comment (yashykt) [reply test/cpp/end2end/xds/xds_end2end_test.cc:7541]: Yes, it does :)
With this change, we were suddenly able to just reuse existing connections when there was a change in certificates. So, those existing connections were just being reused. The only time new certificates would be used are when the subchannel needs to connect again, and hence the need to restart the servers to force a reconnect

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 7 above)
Comment (markdroth) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:129]: Sure, but we have no idea when we'll upgrade to C++20, so that doesn't solve the problem for now.  Between now and then, I don't want to be in a situation where forgetting to add a field will cause a security problem.  I think we need some automation to address this.

For example, maybe we can write a small script to generate this file from input that defines each field name and type, so it can automatically generate the comparator method.  The script could also auto-generate the test to verify that the comparison fails when each field's value differs.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 6 above)
Comment (markdroth) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:138]: Okay, then let's do the same thing for the provider.  I'd prefer to be consistent here.  Just because we don't need it yet doesn't mean the infrastructure should not be in place for it.

Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:138]: Added. The implementation of those comparators might have consequences that I don't want to deal with here, so leaving those as TODOs.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 7 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:134]: Done

File: test/core/security/credentials_test.cc (hunk 11 above)
Comment (yashykt) [reply test/core/security/credentials_test.cc:3708]: Done

**Review by EastyWeasty — APPROVED** [review]

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h [hunk 16]
```diff
@@ -46,10 +47,43 @@
...
+  // `grpc_tls_certificate_provider` objects when they are present in
+  // channel_args. One important usage of this is when channel args are used in
+  // SubchannelKey, which leads to a useful property that allows subchannels to
+  // be reused when two different `grpc_tls_certificate_provider` objects are
+  // used but they compare as equal (assuming other channel args match).
+  int cmp(const grpc_tls_certificate_provider* other) const {
+    GPR_ASSERT(other != nullptr);
+    // Intentionally uses grpc_core::QsortCompare instead of strcmp as a safety
+    // against different grpc_tls_certificate_provider types using the same
+    // name.
```
Comment (ZhenLian) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h:73]: Sorry, I don't get the benefit of using `grpc_core::QsortCompare` over `strcmp`.  If the names are the same, I think they will both return 0?

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc [hunk 17]
```diff
@@ -30,9 +30,12 @@
 
 namespace grpc_core {
 
+const char StaticDataCertificateProvider::kType[] = "static";
```
Comment (ZhenLian) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc:33]: Suggest calling this "static_data"..

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc [hunk 18]
```diff
@@ -94,10 +97,13 @@ gpr_timespec TimeoutSecondsToDeadline(int64_t seconds) {
 
 }  // namespace
 
+const char FileWatcherCertificateProvider::kType[] = "file";
```
Comment (ZhenLian) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc:100]: Suggest calling this "file_watcher"

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 7 above)
Comment (ZhenLian) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:134]: Wow the script is super cool. Thanks for doing that!

**Review by ZhenLian — APPROVED** [review]
LGTM for all the TLS credential related changes. I left some minor questions though.
Thanks for making the change!

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h (hunk 16 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h:73]: So, the worry here is that what if by mistake, we use the same name for two different implementations? In that case, we would hope that the implementations atleast use different definitions of that string. We are just being defensive here.

File: src/core/ext/xds/certificate_provider_store.h [hunk 19]
```diff
@@ -83,9 +84,17 @@ class CertificateProviderStore
       return certificate_provider_->interested_parties();
     }
 
+    int cmp_impl(const grpc_tls_certificate_provider* other) const override {
```
Comment (markdroth) [inline src/core/ext/xds/certificate_provider_store.h:87]: In this case, I think it should delegate to the `Compare()` method of the wrapped provider.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 20]
```diff
@@ -57,19 +61,20 @@ struct grpc_tls_certificate_verifier
...
-  int cmp(const grpc_tls_certificate_verifier* other) const {
+  int Compare(const grpc_tls_certificate_verifier* other) const {
     GPR_ASSERT(other != nullptr);
     int r = strcmp(type(), other->type());
     if (r != 0) return r;
-    return cmp_impl(other);
+    return CompareImpl(other);
   }
 
   const char* type() const { return type_; }
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:71]: As we've discussed offline, let's change the API for all of the classes involved here (credentials, verifier, provider, and anything else) such that instead of passing the type name pointer to the base class and requiring the base class to store it, we instead just make the `type()` method pure virtual, so that we can directly inline the string in that method in the subclasses.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h (hunk 4 above)
Comment (markdroth) [reply src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:45]: Please update these comments with the same wording change as per my latest comments in #28973.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h (hunk 7 above)
Comment (markdroth) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h:134]: This looks great!

Please also add these generated files to the [`.gitattributes`](https://github.com/grpc/grpc/blob/master/.gitattributes) file so that they don't show up in diffs by default.

Also, please make sure this script is automatically run as part of the sanity check before PR submission, so that people don't accidentally manually edit these files and not notice that they actually need to do it via the script.

**Review by markdroth — COMMENTED** [review]
This is moving in the right direction!

Please let me know if you have any questions.  Thanks!

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h (hunk 4 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:45]: Done

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h (hunk 20 above)
Comment (yashykt) [reply src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:71]: Done

File: src/core/ext/xds/certificate_provider_store.h (hunk 19 above)
Comment (yashykt) [reply src/core/ext/xds/certificate_provider_store.h]: Modified the TODO. (I would really prefer to mess with these in this PR.)

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h [hunk 21]
```diff
@@ -34,6 +35,12 @@
 #include "src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h"
 #include "src/core/lib/security/security_connector/ssl_utils.h"
 
+// Known grpc_tls_certificate_provider implementation types
+#define GRPC_TLS_CERTIFICATE_PROVIDER_TYPE_STATIC_DATA "StaticData"
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h:39]: Similar comment here: I don't think we want these as macros.  We should inline them in the individual provider impls.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h [hunk 22]
```diff
@@ -50,6 +57,36 @@ struct grpc_tls_certificate_provider
...
+
+  // The pointer value \a type is used to uniquely identify a creds
+  // implementation for down-casting purposes. Every provider implementation
+  // should use a unique string instance, which should be returned by all
+  // instances of that provider implementation.
+  virtual const char* type() const = 0;
+
+ private:
+  // Implementation for `Compare` method intended to be overridden by
+  // subclasses. Only invoked if `type()` and `other->type()` compare equal as
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h:87]: s/compare equal as strings/point to the same string/

File: src/core/lib/security/credentials/credentials.h [hunk 23]
```diff
@@ -154,14 +157,16 @@ struct grpc_channel_credentials
...
-  const char* type() const { return type_; }
+  // The pointer value \a type is used to uniquely identify a creds
+  // implementation for down-casting purposes. Every creds implementation should
+  // use a unique string instance, which should be returned by all instances of
+  // that creds implementation.
+  virtual const char* type() const = 0;
 
  private:
   // Implementation for `cmp` method intended to be overridden by subclasses.
   // Only invoked if `type()` and `other->type()` compare equal as strings.
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:168]: s/compare equal as strings/point to the same string/

File: src/core/lib/security/credentials/credentials.h [hunk 24]
```diff
@@ -240,14 +244,17 @@ struct grpc_call_credentials
...
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:255]: s/compare equal as strings/point to the same string/

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 25]
```diff
@@ -33,12 +34,15 @@
 #include "src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h"
 #include "src/core/lib/security/security_connector/ssl_utils.h"
 
+// Known grpc_tls_certificate_verifier types
+#define GRPC_TLS_CERTIFICATE_VERIFIER_TYPE_EXTERNAL "External"
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:38]: Same as above: We should not have these macros here.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 26]
```diff
@@ -52,6 +56,28 @@ struct grpc_tls_certificate_verifier
...
+
+  // The pointer value \a type is used to uniquely identify a verifier
+  // implementation for down-casting purposes. Every verifier implementation
+  // should use a unique string instance, which should be returned by all
+  // instances of that verifier implementation.
+  virtual const char* type() const = 0;
+
+ private:
+  // Implementation for `Compare` method intended to be overridden by
+  // subclasses. Only invoked if `type()` and `other->type()` compare equal as
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:78]: s/compare equal as strings/point to the same string/

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h [hunk 27]
```diff
@@ -52,6 +56,28 @@ struct grpc_tls_certificate_verifier
   // Operations that will be performed when a request is cancelled.
   // This is only needed when in async mode.
   virtual void Cancel(grpc_tls_custom_verification_check_request* request) = 0;
+
+  // Compares this grpc_tls_certificate_verifier object with \a other.
+  // If this method returns 0, it means that gRPC can treat the two certificate
+  // verifiers as effectively the same.
+  int Compare(const grpc_tls_certificate_verifier* other) const {
+    GPR_ASSERT(other != nullptr);
+    int r = strcmp(type(), other->type());
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.h:65]: This should use `QsortCompare()`.

File: src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc [hunk 28]
```diff
@@ -30,6 +30,14 @@
 
 namespace grpc_core {
 
+//
+// ExternalCertificateVerifier
+//
+
+ExternalCertificateVerifier::ExternalCertificateVerifier(
```
Comment (markdroth) [inline src/core/lib/security/credentials/tls/grpc_tls_certificate_verifier.cc:37]: I think this can be inlined in the .h file again now.

File: src/core/lib/security/credentials/xds/xds_credentials.h [hunk 29]
```diff
@@ -23,24 +23,51 @@
...
+              std::function<void(absl::Status)>,
+              absl::Status* sync_status) override;
+  void Cancel(grpc_tls_custom_verification_check_request*) override;
+
+  const char* type() const override {
+    return GRPC_TLS_CERTIFICATE_VERIFIER_TYPE_XDS;
+  }
+
+ private:
+  static const char kType[];
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.h:49]: Don't need this anymore.

File: src/core/lib/security/credentials/xds/xds_credentials.h [hunk 30]
```diff
@@ -23,24 +23,51 @@
...
+
+  int CompareImpl(const grpc_tls_certificate_verifier* other) const override;
+
+  RefCountedPtr<XdsCertificateProvider> xds_certificate_provider_;
+  std::string cluster_name_;
+};
+
 extern const char kCredentialsTypeXds[];
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.h:57]: We don't need this anymore.

File: src/core/lib/security/credentials/xds/xds_credentials.cc [hunk 31]
```diff
@@ -29,7 +29,7 @@
 
 namespace grpc_core {
 
-const char kCredentialsTypeXds[] = "Xds";
+const char kCredentialsTypeXds[] = GRPC_CREDENTIALS_TYPE_XDS;
```
Comment (markdroth) [inline src/core/lib/security/credentials/xds/xds_credentials.cc:32]: Don't need this anymore.

File: tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh [hunk 32]
```diff
@@ -32,7 +32,7 @@ for dir in $DIRS
...
+    files="$files `find ${CLANG_FORMAT_ROOT}/$dir -name $glob -and -not -name '*.generated.*' -and -not -name '*.upb.h' -and -not -name '*.upb.c' -and -not -name '*.upbdefs.h' -and -not -name '*.upbdefs.c' -and -not -name '*.pb.h' -and -not -name '*.pb.c' -and -not -name '*.pb.cc' -and -not -name '*.pbobjc.h' -and -not -name '*.pbobjc.m' -and -not -name '*.pbrpc.h' -and -not -name '*.pbrpc.m' -and -not -name end2end_tests.cc -and -not -name end2end_nosec_tests.cc -and -not -name public_headers_must_be_c89.c -and -not -name grpc_shadow_boringssl.h -and -not -name grpc_tls_credentials_options.h -and -not -name grpc_tls_credentials_options_comparator_test.cc`"
```
Comment (markdroth) [inline tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh:35]: Why don't we want to clang-format these generated files?  Historically, we have done that sort of thing -- e.g., we used to do that with the static metadata file, before it was removed.

File: tools/run_tests/sanity/sanity_tests.yaml [hunk 33]
```diff
@@ -27,6 +27,7 @@
 - script: tools/distrib/check_trailing_newlines.sh
 - script: tools/distrib/check_upb_output.sh
 - script: tools/distrib/check_pytype.sh
+- script: tools/codegen/core/gen_grpc_tls_credentials_options.py
```
Comment (markdroth) [inline tools/run_tests/sanity/sanity_tests.yaml:30]: This script will generate the new files, but will it fail if the files in the commit do not match the generated files?  I think that's what we need here.

Also, please add the script to generate the new files to tools/distrib/sanitize.sh.

**Review by markdroth — COMMENTED** [review]
The overall structure here looks very good; comments are all relatively minor.

Please let me know if you have any questions.  Thanks!

File: src/core/lib/security/credentials/credentials.h [hunk 34]
```diff
@@ -53,11 +53,21 @@ typedef enum {
   "FakeTransportSecurity"
 #define GRPC_CHANNEL_CREDENTIALS_TYPE_GOOGLE_DEFAULT "GoogleDefault"
 #define GRPC_CREDENTIALS_TYPE_INSECURE "insecure"
+#define GRPC_CREDENTIALS_TYPE_COMPOSITE "Composite"
```
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/lib/security/credentials/credentials.h:56]: My reasoning for having them all in a single place is to just make it easier on someone creating a new type to look over all the existing names so that it doesn't create any confusion. But, given that we are doing pointer comparison instead of string comparison, I don't think it matters one way or the other.

File: tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh (hunk 32 above)
Comment (yashykt) [reply tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh:35]: I could invoke clang-format but that usually runs under a docker-file, which would just unnecessarily increase the run-time of the sanity check. Also, given that we are already not formatting other generated files such as those from upb or from our tests, I think its fine for these files to follow the same.

File: tools/run_tests/sanity/sanity_tests.yaml (hunk 33 above)
Comment (yashykt) [reply tools/run_tests/sanity/sanity_tests.yaml]: Ah, you are right. I though the test checked for file differences on its own. Added a test mode to the script.

File: src/core/lib/security/credentials/insecure/insecure_credentials.h [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+      const grpc_channel_args* /* args */) override;
+
+  static const char* Type();
+
+  const char* type() const override { return Type(); }
+};
+
+}  // namespace grpc_core
+
+#endif /* GRPC_CORE_LIB_SECURITY_CREDENTIALS_INSECURE_INSECURE_CREDENTIALS_H \
```
Comment (markdroth) [inline src/core/lib/security/credentials/insecure/insecure_credentials.h:57]: Please use C++-style comments.

File: tools/codegen/core/gen_grpc_tls_credentials_options.py [hunk 36]
```diff
@@ -337,3 +351,27 @@ def put_copyright(f, year):
 }""",
       file=T)
 T.close()
+
+if (test_mode):
+    header_diff = False
+    test_diff = False
+    if not filecmp.cmp(
```
[the remark this answers is not in the document]
Comment (yashykt) [reply tools/codegen/core/gen_grpc_tls_credentials_options.py]: Done!

File: tools/codegen/core/gen_grpc_tls_credentials_options.py [hunk 37]
```diff
@@ -337,3 +351,27 @@ def put_copyright(f, year):
...
+        header_diff = True
+    if not filecmp.cmp(
+            test_file_name,
+            'test/core/security/grpc_tls_credentials_options_comparator_test.cc'
+    ):
+        test_diff = True
+    os.unlink(header_file_name)
+    os.unlink(test_file_name)
+    if header_diff:
+        sys.exit(
```
[the remark this answers is not in the document]
Comment (yashykt) [reply tools/codegen/core/gen_grpc_tls_credentials_options.py]: Done!

File: src/core/ext/filters/client_channel/lb_policy/xds/cds.cc [hunk 38]
```diff
@@ -523,7 +523,7 @@ grpc_error_handle CdsLb::UpdateXdsCertificateProvider(
   grpc_channel_credentials* channel_credentials =
       grpc_channel_credentials_find_in_args(args_);
   if (channel_credentials == nullptr ||
-      channel_credentials->type() != kCredentialsTypeXds) {
+      channel_credentials->type() != XdsServerCredentials::Type()) {
```
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/ext/filters/client_channel/lb_policy/xds/cds.cc]: Done. Thanks!

**Review by markdroth — APPROVED** [review]
Looks great!

**Comment by yashykt** [comment]
Thanks for reviewing!
